### PR TITLE
add external jwt signature validation

### PR DESCRIPTION
**Pull request by tharindu1st**

### CODE REVIEW

File: en/docs/learn/api-security/oauth2/access-token-types/jwt-tokens.md [hunk 1]
```diff
@@ -192,4 +192,65 @@ Follow the instructions below to import the public certificate into the client t
...
+  ```
+  [[apim.jwt.issuer]]
+  name = "<issuer value of jwt>"
+  [apim.jwt.issuer.jwks]
+  url = "<jwks endpoint of oauth provider>"
+  ```
+
+### Transform JWT Claims
+
+By default WSO2 API Manager can convert the JWT claims into desired claims by configuring the claim mapping as per the following steps.
```
Comment (bhathiya) [inline en/docs/learn/api-security/oauth2/access-token-types/jwt-tokens.md:222]:
```suggestion
In order to convert the incoming JWT claims into desired claims, you can configure the claim mapping as below.
```

File: en/docs/learn/api-security/oauth2/access-token-types/jwt-tokens.md [hunk 2]
```diff
@@ -192,4 +192,65 @@ Follow the instructions below to import the public certificate into the client t
...
+
+1. Navigate to the `deployment.toml` under the `<API-M_HOME>/repository/conf` folder.
```
Comment (bhathiya) [inline en/docs/learn/api-security/oauth2/access-token-types/jwt-tokens.md:224]:
```suggestion
1. Navigate to the `deployment.toml` under the `<API-M_HOME>/repository/conf/` folder.
```

File: en/docs/learn/api-security/oauth2/access-token-types/jwt-tokens.md [hunk 3]
```diff
@@ -192,4 +192,65 @@ Follow the instructions below to import the public certificate into the client t
...
+2. Add the claim mapping as following under the `[[apim.jwt.issuer]]`.
+  ```
+  [[apim.jwt.issuer.claim_mapping]]
+  remote_claim = "<remote claim value>"
+  local_claim = "<local claim value to map>"
+  ....
+  ```
+By Default Claim mapping done in as one to one mapping, If claim mapping needs to be done in advance behaviour, Users can write a JWT transformer that can configure to specific token issuer.
+
```
Comment (bhathiya) [inline en/docs/learn/api-security/oauth2/access-token-types/jwt-tokens.md:233]:
```suggestion
By default, the claim mapping is done in a one-to-one manner. But, if the mapping needs to be done in an advanced manner, the users can write custom JWT transformers which can be attached to each token issuer.
```

Comment (bhathiya) [reply en/docs/learn/api-security/oauth2/access-token-types/jwt-tokens.md:233]: @tharindu1st we need to talk more about or link for more details on how to extend this.

File: en/docs/learn/api-security/oauth2/access-token-types/jwt-tokens.md [hunk 4]
```diff
@@ -192,4 +192,65 @@ Follow the instructions below to import the public certificate into the client t
...
+
+### Scope Validation flow
```
Comment (bhathiya) [inline en/docs/learn/api-security/oauth2/access-token-types/jwt-tokens.md:235]:
```suggestion
### Scope validation
```

File: en/docs/learn/api-security/oauth2/access-token-types/jwt-tokens.md [hunk 5]
```diff
@@ -192,4 +192,65 @@ Follow the instructions below to import the public certificate into the client t
...
+
+By default WSO2 API Manager validates the scopes came in JWT in scopes claim. If JWT contains the scopes in a different claim, users need to map the claim into `scope` claim as per [Transform JWT Claims](#transform-jwt-claims).
+
```
Comment (bhathiya) [inline en/docs/learn/api-security/oauth2/access-token-types/jwt-tokens.md:238]:
```suggestion
By default, the WSO2 API Manager validates the scopes coming in the `scopes` claim of the JWT. If the JWT contains the scopes in a different claim, users can map that claim into `scopes` claim as per [Transform JWT Claims](#transform-jwt-claims) section.
```

File: en/docs/learn/api-security/oauth2/access-token-types/jwt-tokens.md [hunk 6]
```diff
@@ -192,4 +192,65 @@ Follow the instructions below to import the public certificate into the client t
...
+
+### Subscription validation flow.
```
Comment (bhathiya) [inline en/docs/learn/api-security/oauth2/access-token-types/jwt-tokens.md:240]:
```suggestion
### Subscription validation
```

File: en/docs/learn/api-security/oauth2/access-token-types/jwt-tokens.md [hunk 7]
```diff
@@ -192,4 +192,65 @@ Follow the instructions below to import the public certificate into the client t
...
+2. Consumer key need to be mapped into the Application by using [Provisioning Out-of-Band OAuth2 Clients]({{base_path}}/learn/api-security/oauth2/provisioning-out-of-band-oauth-clients/#provisioning-out-of-band-oauth2-clients/).
+
+Subscription validation from Key manager for JWT can be enabled by going through the following steps.
+
+1. Navigate to the `deployment.toml` under the `<API-M_HOME>/repository/conf` folder.
+2. Add the **subscription_validation_via_km** configuration under the `[apim.jwt_authenitcation]`.
+  ```
+  [apim.jwt_authenitcation]
+  subscription_validation_via_km = true
+  ```
```
Comment (bhathiya) [inline en/docs/learn/api-security/oauth2/access-token-types/jwt-tokens.md:256]:
```suggestion
By default, the WSO2 API Manager does not validate subscriptions if the JWT is generated by an external OAuth provider. In order to enable subscription validation for such JWTs, the following need to be satisfied.

1. The JWT needs to have the consumer key of the generated OAuth client application. By default, the API Manager expects the consumer key to be in the `azp` claim or the `consumerKey` claim. If the consumer key comes in a different claim than these 2, the user can map that claim to one of the above 2 claims as per [Transform JWT Claims](#transform-jwt-claims) section.

2. The consumer key needs to be mapped to the application. For that, you can use [Provisioning Out-of-Band OAuth2 Clients]({{base_path}}/learn/api-security/oauth2/provisioning-out-of-band-oauth-clients/#provisioning-out-of-band-oauth2-clients/).

If you can't satisfy the above requirements but still need subscription validation, the gateway can be configured to connect to the key manager to get the subscription validation done. You can enable that with the following steps.

1. Navigate to the `deployment.toml` under the `<API-M_HOME>/repository/conf/` folder.
2. Add the **subscription_validation_via_km** configuration under the `[apim.jwt_authenitcation]`.
  ```
  [apim.jwt_authenitcation]
  subscription_validation_via_km = true
  ```
```

**Comment by Mariangela** [comment]
@bhathiya , can you please check if this can be merged now?

**Comment by Mariangela** [comment]
@tharindu1st is this fix valid for 3.2.0 as well? Have you sent a separate fix for this to the master branch?

**Comment by tharindu1st** [comment]
@Mariangela  this is only valid for 3.1.0

**Pull request merged**